### ISSUE TITLE
Standardize Component __init__

### ISSUE THREAD

**Issue by mcflugen** [landlab/landlab#220]
Many of our components take the name of a file (or a file-like object) as an argument to its `__init__` method. This file is intended to be an initialization file from which the component reads input parameters. There are several problems with this (I'm sure there are more):
* The user is required to know the file format
* The user is tied to this file format
* The user must create a file before initializing their component
* Every component has code that reads parameters from a file and converts to a `dict` - violates the _Don't Repeat Yourself_ principle
* If we decide to change input file formats, we have to change every component

To address these issues, we decided to do away with initializing components through input files. Python has easy-to-use and intuitive built-in syntax for passing key-value pairs to functions. Let's use that instead. It will now be the responsibility of the application (driver) to read any input files. I guess the only question now is, do we want all parameters to be passed as keywords or as a single `dict`. The options are,
```python
def __init__(self, params):
```
or
```python
def __init__(self, param1='value', param2='another value'):
```
The second method is nice because it clearly shows default values for the parameters and also provides a list of all the possible parameters the component can use. The corresponding docstring would also make more sense, as each parameter would be listed as a separate item. A downside is that this could lead to long parameters lists and clutter the definition of the `__init__` method.

Some usage examples,
```python
mpd = ModelParameterDictionary(from_file='input.txt')
component = MyComponent(**mpd)
```
or if the input file is YAML
```python
params = yaml.load(input_file)
component = MyComponent(**params)
```
or without an input file at all
```python
component = MyComponent(dt=1., stop=10.)
```

Are people still OK with this? Is it OK if we go with the keyword style? What about regular arguments for `__init__`? Does anyone have a component that requires some regular arguments? Should we insist that everything is passed through keywords?

**Comment by SiccarPoint** [comment]
I thought we had already agreed to migrate definitely to a single-dictionary style at the launch meeting. I had volunteered to do this, but have been slow. Having a hybrid, where we maintain compatibility with the old pass-a-string style, would be trivial. I'm still happy to do this.

I can see that the full written out signature looks superficially appealing, but I think at the end of the day it's both less clear if there are subtleties to the way the params are passed in (e.g., optional ones), and clutters the signature. The user needs already to be forced to RTFM for components; I kind of like the idea that they are forced to by passing a dict, not individual terms.

**Comment by mcflugen** [comment]
Thanks for the reminder. You all with your "memory".

I'm fine with having all the params in a single dictionary. It certainly would get confusing if a component had additional keywords that were not input parameters. My only concern is with documentation. We'll need a standard way to document these parameters and make sure the documentation stays up-to-date. I suppose it could be as additional metadata stored in the class. I'm not sure I'm crazy about that, though. In any case, I don't think it's preventing us from moving forward on this but we do need to keep this issue in mind.

I would rather switch the `__init__` methods to use only dictionaries and not file names. What do you think about adding a method to the `Component` class (maybe called `from_path`) that, given a path to a file, reads a dictionary from the file and calls the `__init__` method? Or create a decorator that adds an optional string argument and, if provided, reads the param dictionary from a file. Something along those lines, anyway. Whatever we do, we need to get the file reading stuff out of the components.

So, in terms of an actual signature, how about,
```python
def __init__(self, grid, params=None):
    params = params or {}
```
We will allow additional keywords but what about additional arguments? I guess we could allow them too, so long as the first argument is a grid.

This is a big enough change that, I suggest, we do this component-by-component, with a separate pull request for each component. That will allow each component author to look things over.

**Comment by SiccarPoint** [comment]
I *think* I agree with all of this, if I'm following your final part right. Remember we now have a goodly number of actual honest-to-god users using components out in the wild, so here's a place where we actually really have to maintain backward compatibility if its at all possible.

**Comment by mcflugen** [comment]
I hear you. It keeps me up at night. Really, it does. I think our user-base is still small enough that we can get away with not maintaining backward compatibility in *some* (maybe even most?) cases. We're still at version 0.1, so I think it's understood by our users that we're still in the initial development phase. At least I hope it is. Once we're at 1.0, then we're tied to maintaining backward compatibility.

This is an important discussion that we, as a team, need to have as soon as possible.

For this particular case, I'm OK with maintaining some form of backward compatibility - so long as we remove the input file reading from the components and issue warning messages when they use the old-style initialization.

**Comment by gregtucker** [comment]
I guess I like Eric's original suggestion best: have named parameters (ideally with defaults), but best practice is send them as a dictionary. I love that you can send all params as a dict by simply doing

compo = MyComponent(**my_dict)

The advantages of this solution:
1 - helps maintain backward compatibility (though I also agree that we're early enough in the game that we can bend this where needed)
2 - parameters are explicitly listed in docstring and function/method signature
3 - flexibility
4 - but still uniform and standardized enough that, for example, a component can easily behave as a BMI component (i.e., much easier to map BMI.initialize( string ) to Landlab.Component( **dict ) than it would be if Component init differed among components)

**Comment by mcflugen** [comment]
If we do use keywords, we would have to make it clear what parameters are required and which ones are optional. I think, generally, in Python if something is a keyword it is thought to be *optional*. If it weren't, it would just be an argument. In any case, this could be accomplished through the docstring and assigning required parameters a default value like `None` or something else that would be a *bad value*. It would then fall to the component to check for valid parameters and raise an error if things don't check out. I suppose that applies for passing a straight dictionary as well.

**Comment by gregtucker** [comment]
We should try to resolve this soon. I just stumbled on it again while instantiating a LinearDiffuser, which seems to want a ModelParameterDictionary or a file name (string), and doesn't like me giving it a dict...

**Comment by mcflugen** [comment]
For sure.

I was playing with this last week on a few components. I now think we should try to implement this using keywords rather than passing all of the input parameters as a single dictionary. That main reason being is that it's so much clearer what the parameters are and what their default values are (if any). In addition, because we use the numpydoc conventions, there's already the built-in infrastructure for documenting the keywords.

**Comment by gregtucker** [comment]
I agree. If I understand right, we can still use **some_dict to send keywords if we want to. Anyone else have views on this issue?

**Comment by SiccarPoint** [comment]
I don’t like this if you are *forced* to pass the components separately. But if not, whatever (i.e., I can do comp = MyComp(mg, my_dict)). Though it will annoy our users somewhat, just so we all realise that. If I understand correctly, this will stop

comp = MyComp(mg, ‘a_string.txt’)

working entirely (which a pass-a-dict style wouldn’t necessarily, as it’s only a one-line internal test of what got passed). I still favour back compatibility as much as possible. Remember, we don’t have access to our users’ drivers, so can’t make these changes for them. They’ll have to understand a bit more about Python to know how to build a dict to pass. I don’t think that’s necessarily too much, but it is a code-breaking change. Feel like if we’re going to do this, we need to make a single branch with all these code-breaking changes, then commit them in one go. If I was a user, the only thing that would annoy me more than having to change all my drivers, it would be having to do this multiple times.

Input from Jenny might be good on this, as it’s UX related.

**Comment by SiccarPoint** [comment]
Also, I suspect the general quietness indicates not consent, but that people aren’t checking their email or slack on Christmas Eve. We should probably restart this discussion in the New Year…

**Comment by mcflugen** [comment]
What? It's Christmas Eve!? I have to go shopping!

**Comment by SiccarPoint** [comment]
Oooooooops… ;-)

**Comment by mcflugen** [comment]
I don't see keeping backward compatibility as too much of a problem. It can be done through a decorator that can also issue a warning that this style of initialization is deprecated and may be removed in the future.

I suspect most folks will be passing the parameters as `MyComp(grid, **my_dict)`, not individually.

Yes, we should talk about this after the holidays. We should probably have a comparison where we modify a couple components with the different styles to see what people like better.

**Comment by jennyknuth** [comment]
Done playing Santa! Reading my email…and yes would love to discuss more, but maybe not tonight. I think it is best to do it the 'best' way, but would have to really compare and try it to see what that is, how much of an obstacle the change would be to current users, and how much of a permanent improvement it would be for future users. Deprecation warnings seem a reasonable middle ground, but again, it's all a bit new to me and I don't have a strong understanding of the issue and what the standards are (if there are any). I will catch up on this ASAP, but probably not tomorrow… 




—

**Comment by SiccarPoint** [comment]
This is now partially addressed by #256 , but lots more effort still needed!

**Comment by gregtucker** [comment]
Go, Dan, Go!
